src/lasp_core.erl

```diff
@@ -182,20 +182,23 @@ read(Id, Store) ->
 %%      bound.
 %%
 %%      This operation blocks until `Threshold' has been reached.
+%%      As read/6 runs inside a lasp process, we have to block
+%%      outside, or else it may get killed when used in combination
+%%      with other lasp processes (for example, when used as the read
+%%      function of another lasp process).
 %%
 -spec read(id(), value(), store()) -> {ok, var()}.
 read(Id, Threshold, Store) ->
-    Self = self(),
-    ReplyFun = fun({Id1, Type, Metadata, Value}) ->
-                       {ok, {Id1, Type, Metadata, Value}}
-               end,
-    BlockingFun = fun() ->
-                receive
-                    X ->
-                        X
-                end
-            end,
-    read(Id, Threshold, Store, Self, ReplyFun, BlockingFun).
+    ReplyFun = fun
+        ({error, _}) -> error;
+        ({_Id, _Type, _Metadata, _Value}=Found) ->
+            {ok, Found}
+    end,
+    BlockingFun = fun() -> block end,
+    case read(Id, Threshold, Store, self(), ReplyFun, BlockingFun) of
+        {ok, FoundValue} -> {ok, FoundValue};
+        _ -> receive X -> X end
```

Comment:
Yes, it should only `receive` on `block`, I'll make it explicit and add a match for the error